File: packages/ingestion/src/chunking.ts
export type TextChunk = {
  index: number;
  content: string;
  charStart: number;
  charEnd: number;
};

export function chunkText(
  text: string,
  chunkSize: number,
  chunkOverlap: number,
): TextChunk[] {
  const normalized = text.replace(/\r\n/g, "\n").trim();
  if (normalized.length === 0) return [];

  const size = Math.max(200, chunkSize);
  const overlap = Math.min(Math.max(0, chunkOverlap), Math.floor(size / 2));
  const step = Math.max(1, size - overlap);

  const chunks: TextChunk[] = [];
  let start = 0;
  let index = 0;
  while (start < normalized.length) {
    const end = Math.min(normalized.length, start + size);
    const content = normalized.slice(start, end).trim();
    if (content.length > 0) {
      chunks.push({
        index,
        content,
        charStart: start,
        charEnd: end,
      });
      index += 1;
    }
    if (end >= normalized.length) break;
    start += step;
  }
  return chunks;
}

File: rag-service/rag_service/domain/chunking.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class TextChunk:
    index: int
    content: str
    char_start: int
    char_end: int


def chunk_text(text: str, chunk_size: int, chunk_overlap: int) -> list[TextChunk]:
    normalized = text.replace("\r\n", "\n").strip()
    if not normalized:
        return []

    size = max(200, chunk_size)
    overlap = min(max(0, chunk_overlap), size // 2)
    step = max(1, size - overlap)

    chunks: list[TextChunk] = []
    start = 0
    index = 0
    while start < len(normalized):
        end = min(len(normalized), start + size)
        content = normalized[start:end].strip()
        if content:
            chunks.append(
                TextChunk(
                    index=index,
                    content=content,
                    char_start=start,
                    char_end=end,
                )
            )
            index += 1
        if end >= len(normalized):
            break
        start += step
    return chunks
